File: razer-sdk/src/test/java/com/bonepl/razersdk/sdk/RzKeyJoinerTest.java
package com.bonepl.razersdk.sdk;

import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class RzKeyJoinerTest {
    @Test
    void testSimpleJoin() {
        //when
        final List<RzKey> rzKeys = new RzKeyJoiner()
                .with(RzKey.RZKEY_A).with(RzKey.RZKEY_B).with(RzKey.RZKEY_A).join();

        //then
        assertFalse(rzKeys.isEmpty());
        assertEquals(2, rzKeys.size());
        assertTrue(rzKeys.containsAll(List.of(RzKey.RZKEY_A, RzKey.RZKEY_B)));
    }

    @Test
    void testArrayJoin() {
        //when
        final List<RzKey> rzKeys = new RzKeyJoiner()
                .with(RzKey.RZKEY_A, RzKey.RZKEY_B).with(RzKey.RZKEY_A, RzKey.RZKEY_C).join();

        //then
        assertFalse(rzKeys.isEmpty());
        assertEquals(3, rzKeys.size());
        assertTrue(rzKeys.containsAll(List.of(RzKey.RZKEY_A, RzKey.RZKEY_B, RzKey.RZKEY_C)));
    }

    @Test
    void testEmptyJoins() {
        //when
        final List<RzKey> rzKeys = new RzKeyJoiner().with()
                .with(Collections.emptyList())
                .with(new RzKeySelector()).join();

        //then
        assertTrue(rzKeys.isEmpty());
    }

    @Test
    void testCollectionJoin() {
        //when
        final List<RzKey> rzkeys1 = List.of(RzKey.RZKEY_A, RzKey.RZKEY_A);
        final Set<RzKey> rzKeys2 = Collections.singleton(RzKey.RZKEY_B);
        final List<RzKey> rzKeys3 = Collections.singletonList(RzKey.RZKEY_B);

        //when
        final List<RzKey> rzKeys = new RzKeyJoiner()
                .with(rzkeys1).with(rzKeys2).with(rzKeys3).join();

        //then
        assertFalse(rzKeys.isEmpty());
        assertEquals(2, rzKeys.size());
        assertTrue(rzKeys.containsAll(List.of(RzKey.RZKEY_A, RzKey.RZKEY_B)));
    }

    @Test
    void testKeySelectorJoin() {
        //when
        final RzKeySelector rzKeySelector = new RzKeySelector()
                .withRowOf(RzKey.RZKEY_1).withColumnBetween(RzKey.RZKEY_1, RzKey.RZKEY_2);
        final RzKeySelector rzKeySelector2 = new RzKeySelector()
                .withRowBetween(RzKey.RZKEY_1, RzKey.RZKEY_Q).withColumnOf(RzKey.RZKEY_1);

        //then
        final List<RzKey> rzKeys = new RzKeyJoiner()
                .with(rzKeySelector).with(rzKeySelector2).join();

        //then
        //then
        assertFalse(rzKeys.isEmpty());
        assertEquals(3, rzKeys.size());
        assertTrue(rzKeys.containsAll(List.of(RzKey.RZKEY_1, RzKey.RZKEY_2, RzKey.RZKEY_Q)));
    }
}

File: test/razer-sdk/sdk/RzKeyJoiner.test.ts
import {describe, expect, test} from 'vitest';
import {RzKeyJoiner} from '../../../src/razer-sdk/sdk/RzKeyJoiner.js';
import {RzKeySelector} from '../../../src/razer-sdk/sdk/RzKeySelector.js';
import {RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

describe('RzKeyJoinerTest', () => {
    test('testSimpleJoin', () => {
        //when
        const rzKeys = new RzKeyJoiner()
            .withKeys(RzKey.RZKEY_A).withKeys(RzKey.RZKEY_B).withKeys(RzKey.RZKEY_A).join();

        //then
        expect(rzKeys.size).not.toBe(0);
        expect(rzKeys.size).toBe(2);
        expect(rzKeys.has(RzKey.RZKEY_A)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_B)).toBe(true);
    });

    test('testArrayJoin', () => {
        //when
        const rzKeys = new RzKeyJoiner()
            .withKeys(RzKey.RZKEY_A, RzKey.RZKEY_B).withKeys(RzKey.RZKEY_A, RzKey.RZKEY_C).join();

        //then
        expect(rzKeys.size).not.toBe(0);
        expect(rzKeys.size).toBe(3);
        expect(rzKeys.has(RzKey.RZKEY_A)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_B)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_C)).toBe(true);
    });

    test('testEmptyJoins', () => {
        //when
        const rzKeys = new RzKeyJoiner().withKeys()
            .withCollection([])
            .withCollection(new Set<RzKey>()).join();

        //then
        expect(rzKeys.size).toBe(0);
    });

    test('testCollectionJoin', () => {
        //when
        const rzkeys1 = [RzKey.RZKEY_A, RzKey.RZKEY_A];
        const rzKeys2 = new Set([RzKey.RZKEY_B]);
        const rzKeys3 = [RzKey.RZKEY_B];

        //when
        const rzKeys = new RzKeyJoiner()
            .withCollection(rzkeys1).withCollection(rzKeys2).withCollection(rzKeys3).join();

        //then
        expect(rzKeys.size).not.toBe(0);
        expect(rzKeys.size).toBe(2);
        expect(rzKeys.has(RzKey.RZKEY_A)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_B)).toBe(true);
    });

    test('testKeySelectorJoin', () => {
        //when
        const rzKeySelector = new RzKeySelector()
            .withRowOf(RzKey.RZKEY_1).withColumnBetween(RzKey.RZKEY_1, RzKey.RZKEY_2);
        const rzKeySelector2 = new RzKeySelector()
            .withRowBetween(RzKey.RZKEY_1, RzKey.RZKEY_Q).withColumnOf(RzKey.RZKEY_1);

        //then
        const rzKeys = new RzKeyJoiner()
            .withSelector(rzKeySelector).withSelector(rzKeySelector2).join();

        //then
        expect(rzKeys.size).not.toBe(0);
        expect(rzKeys.size).toBe(3);
        expect(rzKeys.has(RzKey.RZKEY_1)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_2)).toBe(true);
        expect(rzKeys.has(RzKey.RZKEY_Q)).toBe(true);
    });
});
